File: src/main/kotlin/coraythan/keyswap/cards/CardFilters.kt
package coraythan.keyswap.cards

import coraythan.keyswap.House
import org.springframework.data.domain.Sort

data class CardFilters(
        val includeMavericks: Boolean = false,
        val rarities: Set<Rarity> = setOf(),
        val types: Set<CardType> = setOf(),
        val houses: Set<House> = setOf(),
        val ambers: Set<Int> = setOf(),
        val powers: Set<Int> = setOf(),
        val armors: Set<Int> = setOf(),

        val title: String = "",
        val description: String = "",

        val sort: CardSortOptions = CardSortOptions.SET_NUMBER,
        val sortDirection: SortDirection = SortDirection.ASC
)

enum class CardSortOptions {
    SET_NUMBER,
    CARD_NAME,
    AMBER,
    POWER,
    ARMOR
}

enum class SortDirection(val direction: Sort.Direction) {
    ASC(Sort.Direction.ASC),
    DESC(Sort.Direction.DESC)
}

File: keyswap-ui/src/cards/CardFilters.ts
import { observable } from "mobx"
import * as React from "react"
import { House } from "../houses/House"
import { CardType } from "./CardType"
import { Rarity } from "./rarity/Rarity"

export class CardFilters {
    @observable
    title = ""
    @observable
    description = ""
    rarities: Rarity[] = []
    types: CardType[] = []
    houses: House[] = []
    powers: number[] = []
    ambers: number[] = []
    armors: number[] = []
    sort?: string
    @observable
    sortDirection: "ASC" | "DESC" = "ASC"

    handleTitleUpdate = (event: React.ChangeEvent<HTMLInputElement>) => this.title = event.target.value
    handleDescriptionUpdate = (event: React.ChangeEvent<HTMLInputElement>) => this.description = event.target.value
}
